docs/internal/types.md

```diff
@@ -0,0 +1,359 @@
+
+Capabilities offer language and runtime guarantees:
+ * `iso`, or `isolated`, means this is a pointer to an the sentinel object in a region.
+   Every region has a single sentinel that dominates all of the live objects in the region.
+   Exactly one `iso` pointer exists for any region and the behaviour accessing any mutable object must hold the `iso` pointer for the sentinel object of the corresponding region.
+ * `mut`, or `mutable`, means this object can be modified.
+   All mutable objects are contained by a region, the behaviour modifying the object must own the `iso` pointer that refers to the sentinel object of the region.
+ * `imm`, or `immutable`, means no one can write to, so anyone can read from at any time.
+   You can make `iso` regions immutable but not the other way round, as there's no guarantees who's still reading from it.
+ * `readonly` defines local immutability.
+   No object reached via a `readonly` capability may be modified via that the object is not guaranteed to be immutable.
+   `mut` and `readonly` capabilities to the same object may be visible at the same time.
+
+Since capabilities, unions and intersections are types themselves, an `(A & imm)` is completely different from `(A & mut)` or `(A & iso)`.
+Specifically, you cannot convert `mut` to `imm` (as you would with a `const` cast in C++), or vice versa.
+A `freeze` operation consumes an `iso`, destroying the only copy of the capability that is required to mutate any object in a region, and returns an `imm` capability.
+There is no corresponding `unfreeze` operation to consume an `imm` capability and provide any kind of mutable capability.
+Most importantly, capabilities are not type traits like C++'s `const int * const`, they have specific monotonic semantics that change the set of operations that are permitted with the resulting capability.
+
+## Module / Class model
+
+TODO
+
+## Classes and Interfaces
+
+Classes and interfaces have a similar structure:
+ * Fields: elements of a specific type that have a runtime space allocated.
+ * Methods: functions that belong to a type.
+
+The declaration is trivial:
+```ts
+// Helps receivers know they can call a function named increment
+interface Example
+{
+  // All derived classes will have the common field
+  var common : I64;
+
+  // Classes that implement this interface must implement this method and provide the field
+  // Classes that inherit from this interface must implement this method but get the field from the base class
+  increment();
+};
+
+// This class inherits, so it gets the field `common` from the interface
+class MyType : Example
+{
+  let field : I32;
+
+  // Some other method
+  innerSum(x : I32) : I32
+  {
+    let res : I32 = x + field;
+    res
+  }
+
+  // Implements `increment`
+  increment()
+  {
+    self.common += 1;
+  }
+};
+
+// This class implements the interface, so it must provide the `common` field
```

Comment:
```suggestion
// This class aims to implement the interface without explicit inheritance, so it must provide the `common` field
```

Do we have a syntax to declare the _intent to ascribe_ without requesting field and method inheritance?